### PR TITLE
Implement devEnvironments.annotations to provide annotations to all workspace pods

### PR DESCRIPTION
**Pull request by dkwon17**
This PR is a draft PR for now because it requires the latest unreleased version of the devworkspace-operator

### What does this PR do?
This PR implements the `devEnvironments.annotations` field in the CheCluster CR to provide annotations to all workspace pods

### Screenshot/screencast of this PR

image

### What issues does this PR fix or reference?

Related to:
https://issues.redhat.com/browse/CRW-6550
https://github.com/eclipse-che/che/issues/22809

### How to test this PR?

1. Prepare a patch file with an annotation `hello: world`:
```bash
cat > /tmp/cr-patch.yaml <<EOF
apiVersion: org.eclipse.che/v2
kind: CheCluster
spec:
  devEnvironments:
    annotations:
      hello: world
EOF
```

2. Deploy the operator:
 
#### OpenShift
```bash
./build/scripts/olm/test-catalog-from-sources.sh --cr-patch-yaml /tmp/cr-patch.yaml
```

#### on Minikube

```bash
./build/scripts/minikube-tests/test-operator-from-sources.sh --cr-patch-yaml /tmp/cr-patch.yaml
```

3. Start a workspace and verify that the workspace pod has the `hello: world` annotation
4. Remove the `devEnvironments.annotations` field in the CheCluster CR and restart the workspace.
5. Verify that the `hello: world` annotation is removed
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Development resource are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-development-resources)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/dev-workspace-config/dev_workspace_config_test.go [hunk 1]
```diff
@@ -2266,3 +2266,95 @@ func TestReconcileDevWorkspaceImagePullPolicy(t *testing.T) {
...
+			infrastructure.InitializeForTesting(infrastructure.OpenShiftv4)
+
+			devWorkspaceConfigReconciler := NewDevWorkspaceConfigReconciler()
+			_, _, err := devWorkspaceConfigReconciler.Reconcile(deployContext)
+			assert.NoError(t, err)
+
+			dwoc := &controllerv1alpha1.DevWorkspaceOperatorConfig{}
+			err = deployContext.ClusterAPI.Client.Get(context.TODO(), types.NamespacedName{Name: devWorkspaceConfigName, Namespace: testCase.cheCluster.Namespace}, dwoc)
+			assert.NoError(t, err)
+			assert.Equal(t, testCase.expectedOperatorConfig.Workspace.ImagePullPolicy, dwoc.Config.Workspace.ImagePullPolicy)
```
Comment (AObuchow) [inline pkg/deploy/dev-workspace-config/dev_workspace_config_test.go:2357]: I think this was intended to check the config annotations?
```go
assert.Equal(t, testCase.expectedOperatorConfig.Workspace.podAnnotations, dwoc.Config.Workspace.podAnnotations)
```

Comment (dkwon17) [reply pkg/deploy/dev-workspace-config/dev_workspace_config_test.go]: Thank you, fixed

**Review by ibuziuk — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -5,7 +5,7 @@ go 1.19
 require (
 	github.com/che-incubator/kubernetes-image-puller-operator v0.0.0-20210929175054-0128446f5af7
 	github.com/devfile/api/v2 v2.2.2
-	github.com/devfile/devworkspace-operator v0.25.0
+	github.com/devfile/devworkspace-operator v0.29.0
```
Comment (tolusha) [inline go.mod:8]: Please update DEPENDENCIES.md file with a new DWO version and link to https://clearlydefined.io/

File: go.mod [hunk 3]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/sirupsen/logrus v1.8.1
 	github.com/stretchr/testify v1.8.4
 	go.uber.org/zap v1.24.0
-	golang.org/x/net v0.19.0
+	golang.org/x/net v0.23.0
```
Comment (tolusha) [inline go.mod:17]: Please update DEPENDENCIES.md file to update the version of the following libs (no links to  https://clearlydefined.io/ are requested) 
- golang.org/x/net v0.23.0
- golang.org/x/sys v0.18.0
- golang.org/x/term v0.18.0

File: api/v2/checluster_types.go [hunk 4]
```diff
@@ -189,6 +189,9 @@ type CheClusterDevEnvironments struct {
 	// +optional
 	// +kubebuilder:validation:Enum=Always;IfNotPresent;Never
 	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	// Annotations defines additional annotations for workspace pods.
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
```
Comment (tolusha) [inline api/v2/checluster_types.go:194]: What do you think?

```suggestion
	WorkspacesPodAnnotations map[string]string `json:"workspacesPodAnnotations,omitempty"`
```

Comment (dkwon17) [reply api/v2/checluster_types.go]: Sounds good, updated

**Comment by dkwon17** [comment]
Thank you for the review @tolusha , WDYT?

**Review by tolusha — APPROVED** [review]

**Comment by dkwon17** [comment]
I've squashed the commits, merging soon

**Pull request merged**